File: 1_Django/horas/models/registro.py
from django.db import models
from .usuario import CustomUser
from .hora import Hora

class Registro(models.Model):
    fecha = models.DateField()
    hora_ingreso = models.TimeField()
    hora_salida = models.TimeField()
    ubicacion = models.CharField(max_length=100)
    usuario = models.ForeignKey(CustomUser, on_delete=models.CASCADE, related_name='registros')
    num_registro = models.IntegerField(unique=True)
    cantidad_horas_extra = models.DecimalField(max_digits=5, decimal_places=2)
    justificacion_hora_extra = models.TextField()
    estado = models.CharField(max_length=50)
    hora = models.ForeignKey(Hora, on_delete=models.SET_NULL, null=True) 

File: 2_Node/backend/models/Registro.js
const { DataTypes } = require('sequelize');
const sequelize = require('../configDb/db').sequelize;

// Importar modelo Hora
const Hora = require('./Hora');

// Tabla principal
const Registro = sequelize.define('Registro', {
  fecha: {
    type: DataTypes.DATEONLY,
    allowNull: false,
  },
  horaIngreso: {
    type: DataTypes.TIME,
    allowNull: false,
  },
  horaSalida: {
    type: DataTypes.TIME,
    allowNull: false,
  },
  ubicacion: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  usuario: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  numRegistro: {
    type: DataTypes.STRING,
    unique: true,
    allowNull: false,
  },
  cantidadHorasExtra: {
    type: DataTypes.FLOAT,
    allowNull: false,
  },
  justificacionHoraExtra: {
    type: DataTypes.TEXT,
    allowNull: true,
  },
  estado: {
    type: DataTypes.STRING,
    defaultValue: 'pendiente',
    allowNull: false,
  },
});

// Tabla intermedia
const RegistroHora = sequelize.define('RegistroHora', {
  cantidad: {
    type: DataTypes.FLOAT,
    allowNull: false,
  },
});

// Asociaciones
Registro.belongsToMany(Hora, {
  through: RegistroHora,
  foreignKey: 'registroId',
});

Hora.belongsToMany(Registro, {
  through: RegistroHora,
  foreignKey: 'horaId',
});

module.exports = Registro;
